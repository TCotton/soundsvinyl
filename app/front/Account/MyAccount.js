import React, { Component } from 'react';
import axios from 'axios';
import Cookies from 'universal-cookie';
import moment from 'moment';
import { homeURI } from '../../helper_constants';
import { createUsername } from '../../helper_functions';
import './account.scss';
import MyAccountForm from './MyAccountForm';

class MyAccount extends Component {

	constructor (props) {
		super(props);

		this.state = {
			loginName: '',
			loginPassword: '',
			registerName: '',
			registerPassword: '',
			error: null
		};

		this.handleSubmitLogin = this.handleSubmitLogin.bind(this);
		this.handleSubmitRegister = this.handleSubmitRegister.bind(this);
		this.handleInputChange = this.handleInputChange.bind(this);
	}

	handleSubmitRegister (e) {
		e.preventDefault();

		const {
			registerName,
			registerPassword
		} = this.state;

		if (registerName && registerPassword) {

			const email = registerName;
			const password = registerPassword;

			this.setState({error: null}); // place this in a lifecycle hook
			const username = createUsername(email);

			axios.post(`${homeURI}/apiV1/user/add`, {email, password, username})
				.then(res => {

					if (res.data.error) {
						this.setState({error: res.data.error});
					}

					if (res.data.auth) {

						const cookies = new Cookies();

						cookies.set('token', res.data.token, {
							expires: moment().add(1, 'days').toDate(),
							path: '/',
							domain: window.location.hostname,
							secure: true,
						});

						// go back to previous page
						window.location.href = document.referrer;
					}

				}).catch((e) => {

				if (e.toString().includes('409')) {
					this.setState({error: 'This email address is already registered'});
				}

			})
		}
	}

	handleSubmitLogin () {
		const { loginName, loginPassword } = this.state;

		if (loginName && loginPassword) {

			this.setState({error: null}); // place this in a lifecycle hook
			// `user/add`

			axios.post(`${homeURI}/apiV1/user/find`, {email: loginName, password: loginPassword})
				.then(res => {

					if (res.data.error) {
						this.setState({error: res.data.error});
					}

					if (res.data.auth) {

						const cookies = new Cookies();

						cookies.set('token', res.data.token, {
							expires: moment().add(1, 'days').toDate(),
							path: '/',
							domain: window.location.hostname,
							secure: true,
						});

						// go back to previous page
						window.location.href = document.referrer;
					}

				});
		}
	}

	handleInputChange (event) {

		const target = event.target;
		const value = target.value;
		const name = target.name;

		this.setState({
			[name]: value
		});

	}

	render () {

		const { loginName, error, loginPassword, registerName, registerPassword } = this.state;

		const errorFunc = error ? () => {
			return (
				<p styleName='error'>
					{error}
				</p>
			)
		} : () => { return null };

		return (
			<main styleName='myAccount'>
				<h2>
					{'My Account'}
				</h2>
				<p>
					{'Either login or register an account with SoundsVinyl from this page'}
				</p>

				{errorFunc()}

				<section>
					<MyAccountForm
						legend='Login'
						name={loginName}
						nameLabel='Your email'
						onFormSubmit={this.handleSubmitLogin}
						onInputChange={this.handleInputChange}
						password={loginPassword}
						passwordLabel='Your password'
					/>

					<form onSubmit={this.handleSubmitRegister}>
						<fieldset>
							<legend>
								{'Register'}
							</legend>

							<label htmlFor='registerName'>
								{'Your email'}
							</label>
							<input
								id='registerName'
								name='registerName'
								onChange={this.handleInputChange}
								required
								type='text'
								value={registerName}
							/>

							<label htmlFor='registerPassword'>
								{'Your password'}
							</label>
							<input
								id='registerPassword'
								name='registerPassword'
								onChange={this.handleInputChange}
								required
								type='password'
								value={registerPassword}
							/>

							<input
								name='registerSubmit'
								type='submit'
								value='Register'
							/>

						</fieldset>
					</form>
				</section>
			</main>
		)
	}
}

export default MyAccount;
