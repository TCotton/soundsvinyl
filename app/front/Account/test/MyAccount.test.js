import React from 'react'
import MyAccount from '../MyAccount'
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import '../../../../enzymeConfig';

describe.only('Component', () => {
	let component
	beforeEach(() => {
		component = renderer.create(<MyAccount />)
	})

	describe('MyAccount', () => {
		it('should be defined', () => {
			expect(component).toBeDefined()
		})

		it('should render correctly', function(){
			const component = renderer.create(<MyAccount />);
			const tree = component.toJSON();
			expect(tree).toMatchSnapshot();
		});

		it('Should capture name correctly in form[0] onChange and change the props accordingly', function(){

			const component = mount(<MyAccount />);
			const input = component.find('form').at(0).find('input').at(0);

			input.instance().value = 'Andy';
			input.simulate('change');
			expect(component.find('form').at(0).find('input').at(0).props().value).toEqual('Andy');
		})

		it('Should capture password correctly in form[0] onChange and change the props accordingly', function(){

			const component = mount(<MyAccount />);
			const input = component.find('form').at(0).find('input').at(1);

			input.instance().value = 'aPassword';
			input.simulate('change');
			expect(component.find('form').at(0).find('input').at(1).props().value).toEqual('aPassword');
		})

		it('Should capture name correctly in form[1] onChange and change the props accordingly', function(){

			const component = mount(<MyAccount />);
			const input = component.find('form').at(1).find('input').at(0);

			input.instance().value = 'Andy';
			input.simulate('change');
			expect(component.find('form').at(1).find('input').at(0).props().value).toEqual('Andy');
		})

		it('Should capture password correctly in form[1] onChange and change the props accordingly', function(){

			const component = mount(<MyAccount />);
			const input = component.find('form').at(1).find('input').at(1);

			input.instance().value = 'aPassword';
			input.simulate('change');
			expect(component.find('form').at(1).find('input').at(1).props().value).toEqual('aPassword');
		})

	})
})
