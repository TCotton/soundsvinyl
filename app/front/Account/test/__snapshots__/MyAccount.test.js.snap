// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component MyAccount should render correctly 1`] = `
<main
  className="account__myAccount___35jgU"
>
  <h2>
    My Account
  </h2>
  <p>
    Either login or register an account with SoundsVinyl from this page
  </p>
  <section>
    <form
      autoComplete="off"
      name="loginForm"
      onSubmit={[Function]}
    >
      <fieldset>
        <legend>
          Login
        </legend>
        <label
          htmlFor="loginFormloginName"
        >
          Your email
        </label>
        <input
          data-autocapitalize="none"
          disabled={false}
          id="loginFormloginName"
          name="loginFormloginName"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        <label
          htmlFor="loginFormloginPassword"
        >
          Your password
        </label>
        <input
          disabled={false}
          id="loginFormloginPassword"
          name="loginFormloginPassword"
          onChange={[Function]}
          required={true}
          type="password"
          value=""
        />
        <input
          name="loginSubmit"
          type="submit"
          value="Login"
        />
      </fieldset>
    </form>
    <form
      autoComplete="off"
      name="registerForm"
      onSubmit={[Function]}
    >
      <fieldset>
        <legend>
          Register
        </legend>
        <label
          htmlFor="registerFormloginName"
        >
          Your email
        </label>
        <input
          data-autocapitalize="none"
          disabled={true}
          id="registerFormloginName"
          name="registerFormloginName"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
        <label
          htmlFor="registerFormloginPassword"
        >
          Your password
        </label>
        <input
          disabled={true}
          id="registerFormloginPassword"
          name="registerFormloginPassword"
          onChange={[Function]}
          required={true}
          type="password"
          value=""
        />
        <input
          name="loginSubmit"
          type="submit"
          value="Login"
        />
      </fieldset>
    </form>
  </section>
</main>
`;
